otherlibs/dune-rpc/private/dune_rpc_private.ml

```diff
@@ -784,6 +836,51 @@ module Client = struct
 
     let disconnected t = Fiber.Ivar.read t.chan.disconnected
 
+    module Subscription = struct
+      type t = Subscription : ('a, 'i) subscription -> t
+
+      let cancel (Subscription subscription) =
+        let* res = Fiber.Ivar.peek subscription.finished in
```

Comment:
Since `finished` seems to be the state we maintain ourselves, we don't need to use Ivar.peek to do it. We can instead use a normal mutable ref. This will remove the weird assumptions this code relies on.
In particular, there's clearly this assumption that the whole block starting with Ivar.peek ending with Ivar.fill is atomic, but e.g. if the monad is Async.Deferred then that's not going to hold.

Reply:
Actually I don't get it: don't we want to fill `finished` when the server terminates the stream (if that ever happens)?

Reply:
> Since finished seems to be the state we maintain ourselves, we don't need to use Ivar.peek to do it. We can instead use a normal mutable ref. This will remove the weird assumptions this code relies on.

How would we wait for the termination of the stream then?

> In particular, there's clearly this assumption that the whole block starting with Ivar.peek ending with Ivar.fill is atomic, but e.g. if the monad is Async.Deferred then that's not going to hold.

If it's not possible to switch to a ref, we can consider adding this primitive to the interface:

```ocaml
val fill_if_empty : 'a t -> 'a -> [`Ok | `Already_full ] fiber
```

> Actually I don't get it: don't we want to fill finished when the server terminates the stream (if that ever happens)?

It should be filled if the subscription is terminated by a server or a client. The purpose being to add cleanup after the subscription is finished or whatever reason.

Reply:
> How would we wait for the termination of the stream then?

You could still use ivar for that, just not use that ivar for peek. Similar how you're doing it on server side, I think.

> val fill_if_empty : 'a t -> 'a -> [`Ok | `Already_full ] fiber

Yeah, that'd be ok.

> It should be filled if the subscription is terminated by a server or a client

What code fills it when it's terminated by the server?